skdownscale/pointwise_models/quantile.py

```diff
@@ -0,0 +1,363 @@
+        if self.detrend:
+            x_qmapped = x_trend.inverse_transform(x_qmapped)
+
+        return x_qmapped
+
+    def _more_tags(self):
+        return {'_xfail_checks': {'check_methods_subset_invariance': 'because'}}
+
+
+class QuantileMappingReressor(RegressorMixin, BaseEstimator):
+
+    _fit_attributes = ['_X_cdf', '_y_cdf']
+
+    def __init__(self, extrapolate=None, n_endpoints=10):
+        self.extrapolate = extrapolate
+        self.n_endpoints = n_endpoints
+
+        if self.n_endpoints < 2:
+            raise ValueError('Invalid number of n_endpoints, must be >= 2')
+
+    def fit(self, X, y, **kwargs):
+        X = check_array(
+            X, dtype='numeric', ensure_min_samples=2 * self.n_endpoints + 1, ensure_2d=True
+        )
+        y = check_array(
+            y, dtype='numeric', ensure_min_samples=2 * self.n_endpoints + 1, ensure_2d=False
+        )
+
+        X = check_max_features(X, n=1)
+
+        self._X_cdf = self._calc_extrapolated_cdf(X, sort=True, extrapolate=self.extrapolate)
+        self._y_cdf = self._calc_extrapolated_cdf(y, sort=True, extrapolate=self.extrapolate)
+
+        return self
+
+    def predict(self, X, **kwargs):
+
+        check_is_fitted(self, self._fit_attributes)
+        X = check_array(X, ensure_2d=True)
+
+        X = X[:, 0]
+
+        sort_inds = np.argsort(X)
+
+        X_cdf = self._calc_extrapolated_cdf(X[sort_inds], sort=False, extrapolate=self.extrapolate)
+
+        left = -np.inf if self.extrapolate in ['min', 'both'] else None
+        right = np.inf if self.extrapolate in ['max', 'both'] else None
+        X_cdf.pp[:] = np.interp(
+            X_cdf.vals, self._X_cdf.vals, self._X_cdf.pp, left=left, right=right
+        )
+
+        # Extrapolate the tails beyond 1.0 to handle "new extremes"
+        if np.isinf(X_cdf.pp).any():
+            lower_inds = np.nonzero(-np.inf == X_cdf.pp)[0]
+            upper_inds = np.nonzero(np.inf == X_cdf.pp)[0]
+            model = LinearRegression()
+            if len(lower_inds):
+                s = slice(lower_inds[-1] + 1, lower_inds[-1] + 1 + self.n_endpoints)
+                model.fit(X_cdf.pp[s].reshape(-1, 1), X_cdf.vals[s].reshape(-1, 1))
+                X_cdf.pp[lower_inds] = model.predict(X_cdf.vals[lower_inds].reshape(-1, 1))
+            if len(upper_inds):
+                s = slice(upper_inds[0] - self.n_endpoints, upper_inds[0])
+                model.fit(X_cdf.pp[s].reshape(-1, 1), X_cdf.vals[s].reshape(-1, 1))
+                X_cdf.pp[upper_inds] = model.predict(X_cdf.vals[upper_inds].reshape(-1, 1))
+
+        # do the full quantile mapping
+        y_hat = np.full_like(X, np.nan)
+        y_hat[sort_inds] = np.interp(X_cdf.pp, self._y_cdf.pp, self._y_cdf.vals)[1:-1]
+
+        # If extrapolate is 1to1, apply the offset between ref and like to the
+        # tails of y_hat
+        if self.extrapolate == '1to1':
+            X_fit_len = len(self._X_cdf.vals)
+            X_fit_min = self._X_cdf.vals[0]
```

Comment:
`self._X_cdf` and `self._y_cdf` are definitely sorted, right? might it be good to do `min`/`max` instead of select based upon indexing? I guess it's faster the indexing way.

Reply:
Yes, they are sorted by definition.